### DISCUSSION TITLE
cfclient: python error "KeyError: 'bcLighthouse4" with grayed out Command Based Control !

### DISCUSSION THREAD

**Discussion by hyperonex** [bitcraze/discussions#1073] [Q&A]
Dear All,
After installing the cfclient. I've been successfully connected to the Crazyflie however the  Command Based Control was grayed out where all the buttons were disabled (non clickable).
While checking the logged messages to the screen I saw the following python error:

`===== 2023.12.06 10:12:37 =====
INFO:cflib.crazyflie:Callback->Connection completed [radio://0/80/2M]
Traceback (most recent call last):
  File "/Users/achraf/Library/Python/3.9/lib/python/site-packages/cfclient/ui/tabs/lighthouse_tab.py", line 391, in _connected
    if self._helper.cf.param.get_value('deck.bcLighthouse4') == '1':
  File "/Users/achraf/Library/Python/3.9/lib/python/site-packages/cflib/crazyflie/param.py", line 381, in get_value
    return self.values[group][name]
KeyError: 'bcLighthouse4'
`
This error was reproduced in macOS Ventura and Windows 10 as well with the same cfclient gui behavior in both OS.
The console was displaying that the crazyflie is ready to fly!

Please note that I didn't connect any input control device.

Any help on that please?
Regards.

**Comment by knmcguire** [comment]
Hi! So I missed it in your support question but did the Crazyflie had a lighthouse deck connected? The command based flight is only active if there is an positioning system attached.

**Reply by hyperonex** [reply]
It has the multiranger and the flow decks mounted! Are they enough as a positioning system?

**Reply by knmcguire** [reply]
The flowdeck is considered a 'positioning system' yes, The crazyflie is able to estimate it's own velocity based on that.

I'll try to reproduce your error on windows 11 with the latest releases

**Reply by knmcguire** [reply]
I don't seem to get the same issue as you with a crazyflie with flowdeck+multiranger as it all works for me, so I would recommend you to do the following:
- install the latest release (2023.11) of cfclient and cflib (`pip uninstall cfclient cflib` and then `pip install cfclient`)
- update the crazyflie firmware (2023.11) via these instructions: https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/#firmware-upgrade

try it again and if the problem still persists, try the same without the multiranger deck.

**Reply by hyperonex** [reply]
Thank you very much!
Actually cflib and cfclient are the latest release as I installed them two days ago. However the firmware is older that the version you mentioned. I'll load the new firmware to them and let you know the result.

**Reply by knmcguire** [reply]
ah yeah that would be interesting to see. I do think we implemented some kind of param fix that might have an influence which requires both the cflib and the firmware to be updated. Would be good to check out.

**Reply by hyperonex** [reply]
Unfortunately the command based flight is still disabled after loading the last firmware and unmounting the multiranger on macOS ventura, Windows 10 and Ubuntu 22.04.

This is the shell output:

`UserWarning: Using SDL2 binaries from pysdl2-dll 2.24.0
INFO:cfclient.gui:Disabling STL printouts
INFO:cfclient.utils.input.inputreaders:Input readers: ['linuxjsdev', 'pysdl2']
INFO:cfclient.utils.input.inputreaders:Could not initialize [linuxjsdev]: Only supported on Linux
INFO:cfclient.utils.input.inputreaders:Successfully initialized [pysdl2]
INFO:cfclient.utils.input.inputinterfaces:Found interfaces: ['leapmotion', 'wiimote', 'zmqpull']
INFO:cfclient.utils.input.inputinterfaces:Could not initialize [leapmotion]: Leap Motion library probably not installed (No module named 'leapsdk')
INFO:cfclient.utils.input.inputinterfaces:Could not initialize [wiimote]: Missing cwiid (wiimote) driver No module named 'cwiid'
INFO:cfclient.utils.config:Dist config read from C:\Users\achra\AppData\Local\Programs\Python\Python311\Lib\site-packages\cfclient/configs/config.json
INFO:cfclient.utils.config:Config file read from [C:\Users\achra\AppData\Local\Bitcraze\cfclient/config.json]
INFO:cfclient.utils.input.inputinterfaces:Could not initialize [zmqpull]: ZMQ input disabled in config file
INFO:cfclient.utils.zmq_led_driver:ZMQ led disabled in config file
INFO:cfclient.utils.zmq_param:ZMQ param disabled in config file
INFO:cfclient.utils.input:Using device blacklist [(VirtualBox|VMware)]
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cflib.crtp.radiodriver:v0.53 dongle with serial N/A found
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for open_toolboxes
INFO:cfclient.ui.tab_toolbox:No config found for open_toolboxes
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cflib.drivers.cfusb:Looking for devices....
INFO:cflib.crtp.prrtdriver:Initialized PRRT driver.
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cflib.crtp.radiodriver:v0.53 dongle with serial N/A found
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cflib.drivers.cfusb:Looking for devices....
INFO:cflib.crtp.prrtdriver:Initialized PRRT driver.
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cflib.crtp.radiodriver:v0.53 dongle with serial N/A found
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cflib.drivers.cfusb:Looking for devices....
INFO:cflib.crtp.prrtdriver:Initialized PRRT driver.
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cflib.crazyflie:Callback->Connection initialized[radio://0/80/2M/E7E7E7E704]
INFO:cflib.crazyflie:We are connected[radio://0/80/2M/E7E7E7E704], request connection setup
INFO:cflib.crazyflie.platformservice:Request _request_protocol_version()
INFO:cflib.crazyflie:Callback->Connected to [radio://0/80/2M/E7E7E7E704]
WARNING:cflib.crazyflie.log:Error no LogEntry to handle id=3
INFO:cflib.crazyflie.platformservice:_crt_service_callback
INFO:cflib.crazyflie.platformservice:Request protocol version
INFO:cflib.crazyflie.platformservice:_platform_callback
INFO:cflib.crazyflie.platformservice:Protocol version (platform): 7
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cflib.crazyflie.toccache:Saved cache to [C:\Users\achra\AppData\Local\Bitcraze\cfclient/cache/C931D955.json]
INFO:cflib.crazyflie:Log TOC finished updating
INFO:cflib.crazyflie.mem:4 memories found
INFO:cflib.crazyflie:Memories finished updating
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cflib.crazyflie.toccache:Saved cache to [C:\Users\achra\AppData\Local\Bitcraze\cfclient/cache/67270310.json]
INFO:cflib.crazyflie:Param TOC finished updating
INFO:cflib.crazyflie:Callback->Connection setup finished [radio://0/80/2M/E7E7E7E704]
INFO:cfclient.utils.logconfigreader:Parsing [stabilizer.json]
INFO:cflib.crazyflie.log:Have successfully started logging for id=1
INFO:cflib.crazyflie.log:Have successfully started logging for id=3
INFO:cflib.crazyflie.log:Have successfully started logging for id=4
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices
INFO:cflib.crazyflie:All parameters updated
INFO:cflib.crazyflie:Callback->Connection completed [radio://0/80/2M/E7E7E7E704]`

And this is the cfclient output:  

`SYS: I am 0x203934305333500600400032 and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_CORE: 0 deck(s) found
IMU: BMI088: Using I2C interface.
IMU: BMI088 Gyro connection [OK].
IMU: BMI088 Accel connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Complementary (1) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
SYS: About to run tests in system.c.
SYS: NRF51 version: 2023.11
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
IMU: BMI088 gyro self-test [OK]
SYS: Self test passed!
STAB: Wait for sensor calibration...
SYS: Free heap: 20496 bytes
STAB: Starting stabilizer loop
SUP: Can not fly
SUP: Ready to fly`

**Reply by knmcguire** [reply]
There is no flowdeck recognized on your crazyflie so that indicates a different issue. Did you try removing the multiranger and trying again? The console input needs to say that it is recognizing the flow deck.

**Reply by hyperonex** [reply]
Thank you very much. Yes indeed the decks was not well mounted!
After upgrading to the last firmware and remounting the decks everything is working well.

**Reply by knmcguire** [reply]
good to hear! happy to have helped 😄
